File: Backend/src/main/java/com/backend/backend/service/interfaces/EventService.java
package com.backend.backend.service.interfaces;

import java.time.LocalDateTime;
import java.util.List;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import com.backend.backend.dto.request.EventRequest;
import com.backend.backend.dto.response.EventResponse;
import com.backend.backend.dto.response.EventStatsResponse;
import com.backend.backend.model.EventStatus;

public interface EventService {
    // Core Event Operations
    EventResponse createEvent(EventRequest request, String organizationId);
    EventResponse updateEvent(String id, EventRequest request);
    void deleteEvent(String id);
    EventResponse getEvent(String id);
    Page<EventResponse> getAllEvents(Pageable pageable);
    
    // Event Status Management
    EventResponse publishEvent(String id);
    EventResponse cancelEvent(String id);
    EventResponse postponeEvent(String id, LocalDateTime newDateTime);
    
    // Event Registration
    void registerVolunteer(String eventId, String volunteerId);
    void unregisterVolunteer(String eventId, String volunteerId);
    List<EventResponse> getVolunteerEvents(String volunteerId, boolean includeHistory);
    List<EventResponse> getOrganizationEvents(String organizationId, boolean includeHistory);
    
    // Event Search
    Page<EventResponse> searchEvents(
        String query,
        List<String> categories,
        String location,
        Double radius,
        List<EventStatus> status,
        Boolean pendingApproval,
        Boolean isDraft,
        Boolean isOwner,
        Boolean hasRegistrations,
        Boolean isRegistered,
        Boolean skillMatch,
        Boolean availableSpots,
        Boolean registrationOpen,
        String userId,
        Pageable pageable
    );
    List<EventResponse> getUpcomingEvents(int days);
    
    // Event Statistics
    EventStatsResponse getEventStats(String eventId);
    EventStatsResponse getOrganizationEventStats(String organizationId);
    
    // Event Notifications
    void scheduleEventReminders();
    void notifyEventParticipants(String eventId, String message);

    List<String> getCategories();
} 

File: Frontend/src/app/feature/event/services/event.service.ts
import { Injectable } from '@angular/core';
import { ApiService } from '../../../core/services/api.service';
import { Observable } from 'rxjs';
import { Event } from '../../../core/models/event.model';

@Injectable({
  providedIn: 'root'
})
export class EventService {
  private readonly baseUrl = '/events';

  constructor(private apiService: ApiService) {}

  createEvent(event: Partial<Event>): Observable<Event> {
    return this.apiService.post<Event>(this.baseUrl, event);
  }

  getEvent(id: string): Observable<Event> {
    return this.apiService.get<Event>(`${this.baseUrl}/${id}`);
  }

  getAllEvents(): Observable<Event[]> {
    return this.apiService.get<Event[]>(this.baseUrl);
  }

  updateEvent(id: string, event: Partial<Event>): Observable<Event> {
    return this.apiService.put<Event>(`${this.baseUrl}/${id}`, event);
  }

  deleteEvent(id: string): Observable<void> {
    return this.apiService.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchEvents(params: { location?: string; skills?: string[]; radius?: number }): Observable<Event[]> {
    return this.apiService.get<Event[]>(`${this.baseUrl}/search`, params);
  }
} 